src/EditorFeatures/CSharpTest/RemoveUnnecessaryParentheses/RemoveUnnecessaryParenthesesTests.cs

```diff
@@ -0,0 +1,815 @@
+        {
+            await TestAsync(
+@"class C
+{
+    void M()
+    {
+        i = $$(1 + 2);
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        i = 1 + 2;
+    }
+}", requireAllParenthesesForClarity: true);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestNestedParenthesizedExpression_TestAvailableWithAlwaysRemove_And_TestAvailableWhenRequiredForClarity()
+        {
+            await TestAsync(
+@"class C
+{
+    void M()
+    {
+        int i = ( $$(1 + 2) );
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        int i = ( 1 + 2 );
+    }
+}", requireAllParenthesesForClarity: true);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestIncrementExpression_TestAvailableWithAlwaysRemove_And_TestAvailableWhenRequiredForClarity()
+        {
+            await TestAsync(
+@"class C
+{
+    void M()
+    {
+        int i = $$(x++);
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        int i = x++;
+    }
+}", requireAllParenthesesForClarity: true);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestLambdaBody_TestAvailableWithAlwaysRemove_And_TestAvailableWhenRequiredForClarity()
+        {
+            await TestAsync(
+@"class C
+{
+    void M()
+    {
+        Func<int> i = () => $$(1);
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        Func<int> i = () => 1;
+    }
+}", requireAllParenthesesForClarity: true);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestArrayElement_TestAvailableWithAlwaysRemove_And_TestAvailableWhenRequiredForClarity()
+        {
+            await TestAsync(
+@"class C
+{
+    void M()
+    {
+        int[] i = new int[] { $$(1) };
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        int[] i = new int[] { 1 };
+    }
+}", requireAllParenthesesForClarity: true);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestWhereClause_TestAvailableWithAlwaysRemove_And_TestAvailableWhenRequiredForClarity()
+        {
+            await TestAsync(
+@"class C
+{
+    void M()
+    {
+        var q = from c in customer
+                where $$(c.Age > 21)
+                select c;
+    }
+}",
+@"class C
+{
+    void M()
+    {
+        var q = from c in customer
+                where c.Age > 21
+                select c;
+    }
+}", requireAllParenthesesForClarity: true);
+        }
+
+        [Fact, Trait(Traits.Feature, Traits.Features.CodeActionsRemoveUnnecessaryParentheses)]
+        public async Task TestCastExpression_TestAvailableWithAlwaysRemove_And_TestAvailableWhenRequiredForClarity()
```

Comment:
why are these names so repetetive and include Test multiple times? as opposed to just _AvailableWithAlwaysRemoveAndWhenRequiredForClarity?

Reply:
Because i found those shorter names too confusing.  :)

Reply:
what about _AvailableWithAlwaysRemove_And_AvailableWhenRequiredForClarity? surely we already know it's a test